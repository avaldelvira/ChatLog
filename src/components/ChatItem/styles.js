import styled, { keyframes } from 'styled-components';

const popUp = keyframes`
  from { transform: scale(0); }
  to { transform: scale(1); }
`;

const slideFromRight = keyframes`
  from { left: 100%; }
  to { left: 0; }
`;

export const Avatar = styled.img`
  width: 3rem;
  min-width: 3rem;
  height: 3rem;
  border-radius: 3rem;
  margin-right: 1rem;
  border: .125rem solid #FFF;
`;

export const ChatMetadata = styled.div`
  font-size: .825rem;
  margin-bottom: .5rem;
  opacity: 0.75;
`;

export const EmailLayer = styled.div`
  position: absolute;
  top:0;
  bottom: 0;
  width: 100%;
  animation-name: ${slideFromRight};
  animation-duration: 300ms;
  animation-fill-mode: forwards;
  background-color: rgba(255,255,255, 0.95);
  color: #223153;
  font-size: 1.125rem;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const ChatItemWrapper = styled.li`
  width:100%;
  display: flex;
  align-items: flex-start;
  border-radius: .5rem;
  background: linear-gradient(-45deg, rgba(2,0,36,1) 0%, rgba(34,49,83,1) 90%);
  padding: 1rem;
  margin-top: 1rem;
  animation-name: ${popUp};
  animation-duration: 300ms;
  animation-delay: ${({index}) => `${index * 50}ms`};
  animation-fill-mode: forwards;
  color: #FFF;
  position: relative;
  overflow: hidden;
  transform: scale(0);
  box-shadow: 0 0 0 0 transparent;
  transition: box-shadow 300ms;
  
  @media screen and (min-width: 756px) {
    width: 48%;
  }
  @media screen and (min-width: 900px) {
    width: 33%;
  }
  &:hover {
    box-shadow: 0 0 2rem 0 rgba(34,49,83,0.25);
  }
 `;


