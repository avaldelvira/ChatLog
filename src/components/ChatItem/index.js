import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { ChatItemWrapper, Avatar, ChatMetadata, EmailLayer } from './styles';

const ChatItem = ({ fullName, timestamp, email, message, index, avatar }) => {
  const [isHovering, setIsHovering] = useState(false);
  const options = {hour: 'numeric', minute: 'numeric'};
  const time = new Date(timestamp).toLocaleDateString('en-GB', options );
  return (
    <ChatItemWrapper
      onMouseOver={() => setIsHovering(true)}
      onMouseLeave={() => setIsHovering(false)}
      index={index}
    >
      {avatar && <Avatar src={avatar} title={email} alt={`${fullName} Avatar`}/>}
      <div>
        <ChatMetadata>{fullName} - {time}</ChatMetadata>
        <div>{message}</div>
      </div>
      {isHovering && <EmailLayer>{email}</EmailLayer>}
    </ChatItemWrapper>
  );
};

ChatItem.propTypes = {
  message: null,
  avatar: null,
};

ChatItem.propTypes = {
  fullName: PropTypes.string.isRequired,
  timestamp: PropTypes.number.isRequired,
  message: PropTypes.string,
  avatar: PropTypes.string,
};

export default ChatItem;