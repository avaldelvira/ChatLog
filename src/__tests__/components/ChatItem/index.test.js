import React from "react";
import { create } from "react-test-renderer";

import ChatItem from '../../../components/ChatItem';

let container;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

const mockMessages = [
  {
    avatar: null,
    email: "abakera@craigslist.org",
    fullName: "Amanda Baker",
    message: "Nulla tempus. Vivamus in felis eu sapien cursus vestibulum. Proin eu mi.",
    messageId: "fa0ca158-42ae-435d-8c32-13f4cb7b8de8",
    timestamp: 1456272088000,
    userId: "73f30d93-f87b-45ea-bda5-60fa4b3716b5",
  },
  {
    avatar: "http://dummyimage.com/100x100.jpg/dddddd/000000",
    email: "awilliams8@indiegogo.com",
    fullName: "Aaron Williams",
    message: "Quisque ut erat. Curabitur gravida nisi at nibh.",
    messageId: "555f4e1b-90a6-4079-850d-54001f29ce44",
    timestamp: 1456481117000,
    userId: "c980934a-f54e-459e-8137-0ec3654cdafb",
  }
];

describe("ChatItem Component", () => {
  test("ChatItem without avatar Matches the snapshot", () => {
    const chatItem = create(<ChatItem {...mockMessages[0]}/>);
    expect(chatItem.toJSON()).toMatchSnapshot();
  });
  test("ChatItem with avatar Matches the snapshot", () => {
    const chatItem = create(<ChatItem {...mockMessages[1]}/>);
    expect(chatItem.toJSON()).toMatchSnapshot();
  });
});
