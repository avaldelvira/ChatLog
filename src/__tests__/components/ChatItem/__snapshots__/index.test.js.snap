exports[`ChatItem Component ChatItem with avatar Matches the snapshot 1`] = `
<li
  className="sc-bxivhb kUeqyn"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}>
  <img
    alt="Aaron Williams Avatar"
    className="sc-bdVaJa kAZrIL"
    src="http://dummyimage.com/100x100.jpg/dddddd/000000"
    title="awilliams8@indiegogo.com" />
  <div>
    <div
      className="sc-bwzfXH JoQuo">
      Aaron Williams
       - 
      2/26/2016, 11:05 AM
    </div>
    <div>
      Quisque ut erat. Curabitur gravida nisi at nibh.
    </div>
  </div>
</li>
`;

exports[`ChatItem Component ChatItem without avatar Matches the snapshot 1`] = `
<li
  className="sc-bxivhb kUeqyn"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}>
  <div>
    <div
      className="sc-bwzfXH JoQuo">
      Amanda Baker
       - 
      2/24/2016, 1:01 AM
    </div>
    <div>
      Nulla tempus. Vivamus in felis eu sapien cursus vestibulum. Proin eu mi.
    </div>
  </div>
</li>
`;
